### DISCUSSION TITLE
How to toggle the grid of a 3D scene?

### DISCUSSION THREAD

**Discussion by WolfgangFahl** [zauberzeug/nicegui#1246] [Q&A] [answered]
### Question

I tried to implement grid toggling:
```python
async def toggle_grid(self,_ea):
        """
        toogle the grid of my scene
        """
        try:
            grid=self.scene._props["grid"]
            grid_str="off" if grid else "on"
            ui.notify(f"setting grid to {grid_str}")
            self.scene._props["grid"]=not grid
            self.scene.update()
        except BaseException as ex:
            self.handleExeption(ex)
        pass
```

but this approach has no effect. I am also having trouble getting scene.clear() to work. The rendering in http://nicescad.bitplan.com/ seems to be unreliable. There might be a bunch of basic things that are not correct. E.g. i think there is no proper client handling in place.

For the toggle of the button's icon issue I'll open a separated dicussion.

**Comment by falkoschindler** [comment, marked as answer]
There is currently no official way to toggle the grid. Here is a workaround (which won't keep the state when refreshing the browser):
```py
ui.button('Show grid', on_click=lambda: ui.run_javascript(f'scene_c{scene.id}.children.find(c => c.type === "GridHelper").visible = true', respond=False))
ui.button('Hide grid', on_click=lambda: ui.run_javascript(f'scene_c{scene.id}.children.find(c => c.type === "GridHelper").visible = false', respond=False))
```

Regarding `element.clear()`: This method keeps confusing users. Its main purpose is to remove child _elements_, not element-specific content like text input or - in case of a 3D scene - 3D _objects_. But in case of `ui.log` we extended `clear()` to also clear the log. And I can imagine to extend `scene.clear()` to also remove all objects.

**Comment by WolfgangFahl** [comment]
see https://github.com/WolfgangFahl/nicescad/issues/20

**grid toggle**
Thanks for the work around I'll look into this.

**scene.clear()**
I imagine to get the clear function work the way the axes helper does it - keep track of added stl files my self and allow to actively clear elements e.g. by clicking and removing them. The .name attribute seems to work nicely and i think keeping an object or id dict by name should not be rocket science.

**Comment by WolfgangFahl** [comment]
The following code works. IMHO it would be good to make the grid toogling part of the ui.scene API.
```python
async def toggle_grid(self,_ea):
        """
        toogle the grid of my scene
        """
        try:
            grid=self.scene._props["grid"]
            grid_str="off" if grid else "on"
            grid_js="false" if grid else "true"
            # try toggling grid
            ui.notify(f"setting grid to {grid_str}")
            grid=not grid
            # workaround according to https://github.com/zauberzeug/nicegui/discussions/1246
            js_cmd=f'scene_c{self.scene.id}.children.find(c => c.type === "GridHelper").visible = {grid_js}'
            await ui.run_javascript(js_cmd, respond=False)
            self.scene._props["grid"]=grid
            self.scene.update()
            # try toggling icon
            self.toggle_icon(self.grid_button)
        except BaseException as ex:
            self.handleExeption(ex)
        pass
```
![grafik](https://github.com/zauberzeug/nicegui/assets/1336221/cfe47172-b825-4bf7-a3c9-8aa3f4b1d5f0)

![grafik](https://github.com/zauberzeug/nicegui/assets/1336221/b5640cd5-5e17-47b1-8db4-973619e0db5b)

**Reply by falkoschindler** [reply]
Indeed. I opened a new feature request #1261.
